smslibrary/src/main/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiver.java

```diff
@@ -53,9 +62,36 @@ public void onReceive(Context context, Intent intent) {
                 break;
         }
 
-        if (listener != null) //extra check, even though listener should never be null
-            listener.onSMSDelivered(message, deliveredState);
-
+        // if delivery of a part failed, gives to the listener its deliveredState containing the
+        // error
+        if (deliveredState != SMSMessage.DeliveredState.MESSAGE_DELIVERED) {
+            listener.onSMSDelivered(reconstructMessage(), deliveredState);
+            context.unregisterReceiver(this);
+            return;
+        }
+        // binary search on messageParts for the SMSPart associated to this intent's action
+        int partIndex = Collections.binarySearch(messageParts, new SMSPart(null, intent.getAction()));
+        if (partIndex >= 0) {
+            SMSPart part = messageParts.get(partIndex);
+            part.setReceived();
+            messageParts.set(partIndex, part);
```

Comment:
messageParts.get(partIndex).setReceived()
maybe it's faster
cause part.setReceived(); dont change the object but it's state

Reply:
You're right, I'll fix this in `SMSSentBroadcastReceiver` too